flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/streamstatus/StatusWatermarkValve.java

```diff
@@ -29,24 +30,13 @@
 /**
  * A {@code StatusWatermarkValve} embodies the logic of how {@link Watermark} and {@link StreamStatus} are propagated to
  * downstream outputs, given a set of one or multiple input channels that continuously receive them. Usages of this
- * class need to define the number of input channels that the valve needs to handle, as well as provide a customized
- * implementation of {@link ValveOutputHandler}, which is called by the valve only when it determines a new watermark or
- * stream status can be propagated.
+ * class need to define the number of input channels that the valve needs to handle, as well as provide a implementation of
+ * {@link DataOutput}, which is called by the valve only when it determines a new watermark or stream status can be propagated.
  */
 @Internal
 public class StatusWatermarkValve {
 
-	/**
-	 * Usages of {@code StatusWatermarkValve} should implement a {@code ValveOutputHandler}
-	 * to handle watermark and stream status outputs from the valve.
-	 */
-	public interface ValveOutputHandler {
-		void handleWatermark(Watermark watermark) throws Exception;
-
-		void handleStreamStatus(StreamStatus streamStatus) throws Exception;
-	}
-
-	private final ValveOutputHandler outputHandler;
+	private final DataOutput output;
```

Comment:
Thanks for this concern. I actually also hesitated to this issue before and passed the `output` in specific methods in `StatusWatermarkValve` instead of constructor in previous version.

I think there are two different possibilities here:

- Make `StatusWatermarkValve` completely coupled with `StreamTaskNetworkInput`, then we could even construct `StatusWatermarkValve` inside constructor of `StreamTaskNetworkInput`. This is a positive point. And we assume that the `output` is same for both `emitNext` and `StatusWatermarkValve` now. But if the `output` instances are different in future as you mentioned, then we need to pass another `output` into `StreamTaskNetworkInput` for `StatusWatermarkValve`. Especially it would cause many changes in `StatusWatermarkValve` because there are four internal methods involved in passing `output` argument in `StatusWatermarkValve`.

- Construct the different `output` instances in the upper layer of `StreamInputProcessor` for both `StreamTaskNetworkInput` and `StatusWatermarkValve` as the current implementation in this PR. In other words `StatusWatermarkValve` could be regarded as a `output` wrapper which is actually used in `StatusWatermarkValve`, and the wrapped `output` inside `StatusWatermarkValve` could be anything else different from the `output` in `emitNext`. If it changes we only need to construct a different `output` instance in `StreamInputProcessor` and pass it into `StatusWatermarkValve`, no need to touch four internal methods inside it.

I have no obvious preference for above two ways, because we could think of side effects for any way.

For the alternative of encapsulating `StatusWatermarkValve` behind `output`, I wonder it is not good to let `output` see any channel index concept, especially considering the source output future. And the current `StatusWatermarkValve` is a bit heavy-weight, so encapsulating it into `output` would complicate `output` implementation. In contrast, making `output` as an field of `StatusWatermarkValve` seems more light-weight.

Reply:
By encapsulating `StatusWatermarkValve` behind the `output`, I don't mean inlining the implementation of `StatusWatermarkValve` in the `output` class, but rather making it a private field (in similar way how you did it now, by moving the valve to the source).

Abstraction wise, I'm not sure (have no strong opinion) whether the valve should belong to either the source or output class.
But current situation, that one of the fields in the source have reference to the output and the contact of `PushingAsyncDataInput` that the output is passed via main `emitNext()` method is poor by design, in my opinion.

> there are four internal methods involved in passing output argument in `StatusWatermarkValve`.

That is not a lot of change, imo.

Reply:
It can not directly work well to make `StatusWatermarkValve` as a private field of `output`. 
The overall process is `StreamTaskNetworkInput#processElement` -> `if (recordOrMark.isWatermark())` -> `output.emitWatermark` -> StatusWatermarkValve#inputWatermark`.

The key problem is that the interface of `output.emitWatermark` can not expose any channel index info, but `StatusWatermarkValve` has to know the current channel index to judge the internal logics. So we make the `StatusWatermarkValve` in the upper layer of `output`. In other words, the `StatusWatermarkValve` can wrap the `output`, but `output` can not wrap the `StatusWatermarkValve`.

As for passing the `output` in the constructor of `StatusWatermarkValve` vs passing it in four internal methods of `StatusWatermarkValve`. I still prefer to the current way for two reasons:

- Once `StatusWatermarkValve` needs a different output future, we only need to change one place, otherwise we need to touch four internal methods.

- From design aspect, I think it is better to reduce the unnecessary dependency between `StatusWatermarkValve` and `StreamTaskNetworkInput`. Currently we have to rely on the channel index from `StreamTaskNetworkInput`. But if we regard the `output` as a more general concept which might not be consistent with the `output` in `emitNext`, then we can pass it into the constructor in free from outside. Just like the previous `ForwardingValveOutputHandler` which was also passed into the constructor of `StatusWatermarkValve` from outside.